Suggest improvements

saefstroem opened the pull request:
Please suggest improvements in alignment with release.

saefstroem committed:
Initial draft of structure

Imported sled database interface from JetRPC

saefstroem committed:
First untested draft

- Builtin erc20 abi used Ethereum Binance Pegged Contract.
- Wrapper struct for erc20 token
- get_balance for erc20 token
- refactored sled db interface
- set gateway callback function to be async
- drafted poller function (untested)
- invoice struct

todo: test the poller by using this as external library and also draft test cases to test the poller to the extent possible without involving private keys into the repo

saefstroem committed:
Changed project structure, docs, mods

- changed project structure to prevent internal interfaces from being exported to the public API.
- added custom expiration for invoices
- added descriptive names for payment gateways

saefstroem committed:
Clippy suggestions

Refactored according to suggestions by clippy.

saefstroem committed:
Added initial readme

saefstroem committed:
Added description

Added description of how the library functions

saefstroem committed:
Ported to alloy, deprecate rust-web3, treasury transfer

- Deprecated rust-web3 in favor of alloy-rs due to rust-web3 being deprecated in the future
- Replaced eth abi with IERC20 abi.
- Transfer paid invoice funds to treasury when paid

TODO: in the near future token transfers also need to be included so that it is possible to configure payments via tokens.

saefstroem committed:
Set treasury address in tests

saefstroem committed:
Create rust.yml

Add basic rust pipeline

biryukovmaxim commented on src/audit/mod.rs:
```diff
@@ -0,0 +1,25 @@
+/// Internal logging function used to log errors.
+/// Can be disabled by setting environment variable ACCEPTEVM to 0
+pub fn log_sync(data: &str) {
+    match env::var("ACCEPTEVM_LOGS") {
```
usually everyone uses RUST_LOG env
where the crate/module can be specified
https://rust-lang-nursery.github.io/rust-cookbook/development_tools/debugging/config_log.html

saefstroem replied:
e889da77a756abf1d3390a98806640e31f58cfb7

biryukovmaxim commented on src/common/mod.rs:
```diff
@@ -0,0 +1,39 @@
+/// Retrieve the current unix time in nanoseconds
+pub fn get_unix_time_seconds() -> u64 {
+    let now = SystemTime::now();
+    let duration = now.duration_since(UNIX_EPOCH).unwrap_or_else(|_| {
+        println!("Failed computing UNIX timestamp during admin login!");
+        Duration::from_secs(0)
+    });
+    duration.as_secs()
```
```suggestion
    let duration = now.duration_since(UNIX_EPOCH).inspect_err(|_| println!("Failed computing UNIX timestamp during admin login!").unwrap_or_default().as_secs()
```

saefstroem replied:
Thanks for this, the print statement can be removed completely. Admin login does not make sense here. I think I copied this from one of my other projects.

saefstroem replied:
e889da77a756abf1d3390a98806640e31f58cfb7

biryukovmaxim commented on src/db/mod.rs:
```diff
@@ -0,0 +1,137 @@
+use sled::Tree;
+
+/// Retrieve a value by key from a tree.
+async fn get_from_tree(db: &Tree, key: &str) -> Result<Vec<u8>, DatabaseError> {
+    match db.get(key) {
+        Ok(result) => match result {
+            Some(value) => Ok(value.to_vec()),
+            None => Err(DatabaseError::NotFound),
+        },
+        Err(_error) => Err(DatabaseError::Get),
+    }
+}
```
check https://docs.rs/thiserror/latest/thiserror/
crate, it can help with error conversion.

then it can be refactored to something like
```rust
async fn get_from_tree(db: &Tree, key: &str) -> Result<Vec<u8>, DatabaseError> {
    db.get(key)?.ok_or(DatabaseError::NotFound)
}
```

saefstroem replied:
Okay thats a good suggestion

saefstroem replied:
e889da77a756abf1d3390a98806640e31f58cfb7

biryukovmaxim commented on src/audit/mod.rs:
```diff
@@ -0,0 +1,25 @@
+
+/// Internal logging function used to log errors.
+/// Can be disabled by setting environment variable ACCEPTEVM to 0
+pub fn log_sync(data: &str) {
```
the name of module is not clear. in `audit` i would like to see some audit reports

saefstroem replied:
e889da77a756abf1d3390a98806640e31f58cfb7

biryukovmaxim commented on src/db/mod.rs:
```diff
@@ -0,0 +1,137 @@
+}
+/// Retrieve all key,value pairs from a specified tree
+async fn get_all_from_tree(db: &Tree) -> Result<Vec<(Vec<u8>, Vec<u8>)>, DatabaseError> {
+    let mut all = Vec::new();
+    for el in db.iter() {
+        match el {
+            Ok(value) => {
+                let el_bin_key = value.0.to_vec();
+                let el_bin_value = value.1.to_vec();
+                all.push((el_bin_key, el_bin_value));
+            }
+            Err(error) => {
+                log_sync(&format!("Db Interaction Error: {}", error));
+                return Err(DatabaseError::Get);
+            }
+        }
+    }
+    Ok(all)
```
```suggestion
    db.iter()
        .map(|res| {
            res.map_err(|error| {
                log_sync(&format!("Db Interaction Error: {}", error));
                DatabaseError::Get
            })
            .map(|(key, value)| (key.to_vec(), value.to_vec()))
        })
        .collect()
```

saefstroem replied:
e889da77a756abf1d3390a98806640e31f58cfb7

biryukovmaxim commented on src/db/mod.rs:
```diff
@@ -0,0 +1,137 @@
+    tree: &sled::Tree,
+) -> Result<Vec<(String, T)>, DatabaseError> {
+    let binary_data = get_all_from_tree(tree).await?;
+    let mut all = Vec::new();
```
```suggestion
    let mut all = Vec::with_capacity(binary_data.len());
```

saefstroem replied:
e889da77a756abf1d3390a98806640e31f58cfb7

biryukovmaxim commented on src/erc20/mod.rs:
```diff
@@ -0,0 +1,64 @@
+
+    /// Retrieves the token balance of a specified address
+    pub async fn get_balance(&self, address: String) -> Result<Uint<256, 4>, Error> {
+        let IERC20::balanceOfReturn { _0 } = self
+            .contract
+            .balanceOf(address.parse().unwrap())
+            .call()
+            .await?;
+        Ok(_0)
```
```suggestion
let IERC20::balanceOfReturn { _0: balance } = self
            .contract
            .balanceOf(address.parse().unwrap())
            .call()
            .await?;
        Ok(balance)
```

saefstroem replied:
e889da77a756abf1d3390a98806640e31f58cfb7

biryukovmaxim committed:
feat: make callback structure mutexless

biryukovmaxim commented on src/poller/mod.rs:
```diff
@@ -0,0 +1,184 @@
+/// Checks if a specific token of a specific amount has been received
+/// at a certain address.
+async fn check_if_token_received(
+    token: ERC20Token,
+    invoice: Invoice,
```
accept by ref, it reduces unnecessary clones

starkbamse replied:
e889da77a756abf1d3390a98806640e31f58cfb7

biryukovmaxim commented on src/poller/mod.rs:
```diff
@@ -0,0 +1,184 @@
+
+/// Retrieves the gas token balance of the specified address on the specified web3 instance
+async fn get_native_balance(
+    provider: RootProvider<Http<Client>>,
+    address: String,
```
accept by ref, it reduces unnecessary clones

saefstroem replied:
e889da77a756abf1d3390a98806640e31f58cfb7

biryukovmaxim commented on src/poller/mod.rs:
```diff
@@ -0,0 +1,184 @@
+
+/// Used to check if the invoice recipient has received enough money to cover the invoice
+async fn check_if_native_received(
+    provider: RootProvider<Http<Client>>,
+    invoice: Invoice,
```
accept by ref, it reduces unnecessary clones

saefstroem replied:
e889da77a756abf1d3390a98806640e31f58cfb7

biryukovmaxim commented on src/poller/mod.rs:
```diff
@@ -0,0 +1,184 @@
+
+/// A function that branches control flow depending on the invoice shall
+/// be paid by an ERC20-compatible token or the native gas token on the network
+async fn check_and_process(provider: RootProvider<Http<Client>>, invoice: Invoice) -> bool {
```
accept by ref, it reduces unnecessary clones

saefstroem replied:
e889da77a756abf1d3390a98806640e31f58cfb7

biryukovmaxim commented on src/types/mod.rs:
```diff
@@ -0,0 +1,54 @@
+    /// Recipient address
+    pub to: String,
+    /// Recipient instance
+    pub wallet: B256,
```
use https://docs.rs/zeroize/latest/zeroize/
it will make it zeroes after drop.
in necessary because if anyone sends more value than expected in case of security issue it can lead to the loosing of funds

saefstroem replied:
e889da77a756abf1d3390a98806640e31f58cfb7

saefstroem committed:
Fix suggestions

- use RUST_LOG
- simplify functions
- convert sled errors to custom db error
- deprecate audit module
- reduce cloning and replace with ref usage
- add zeroize type for wallet bytes

starkbamse commented:
@biryukovmaxim Please review new commits for your requested changes

saefstroem committed:
Merge branch 'main' into make-callback-signature-mutexless

saefstroem committed:
Merge pull request #2 from biryukovmaxim/make-callback-signature-mutexless

feat: make callback structure mutexless

saefstroem committed:
Shrink get_last_from_tree impl

Co-Authored-By: Maxim <59533214+biryukovmaxim@users.noreply.github.com>

saefstroem committed:
Merge branch 'main' of https://github.com/starkbamse/acceptevm
